pkg/broker/ingress/multi_topic_decouple_sink_test.go

```diff
@@ -220,6 +226,224 @@ func TestMultiTopicDecoupleSink(t *testing.T) {
+
+	tests := []struct {
+		name          string
+		brokerTargets map[string]*config.Target
+		hasTrigger    bool
+	}{
+		{
+			name:       "broker with no target",
+			hasTrigger: false,
+		},
+		{
+			name:          "broker with empty target",
+			brokerTargets: map[string]*config.Target{},
+			hasTrigger:    false,
+		},
+		{
+			name: "broker with target with no filters",
+			brokerTargets: map[string]*config.Target{
+				"target_1": {},
+			},
+			hasTrigger: true,
+		},
+		{
+			name: "broker with target with matching filter",
+			brokerTargets: map[string]*config.Target{
+				"target_1": {
+					FilterAttributes: map[string]string{
+						"type":   eventType,
+						"source": "test-source",
+					},
+				},
+			},
+			hasTrigger: true,
+		},
+		{
+			name: "broker with target with non-matching filter",
+			brokerTargets: map[string]*config.Target{
+				"target_1": {
+					FilterAttributes: map[string]string{
+						"type":   eventType,
+						"source": "some-random-source",
+					},
+				},
+			},
+			hasTrigger: false,
+		},
+		{
+			name: "broker with multiple targets with one matching target filter",
+			brokerTargets: map[string]*config.Target{
+				"non_matching_target_1": {
+					FilterAttributes: map[string]string{
+						"type":   eventType,
+						"source": "some-random-source",
+					},
+				},
+				"non_matching_target_2": {
+					FilterAttributes: map[string]string{
+						"source": "some-random-other-source",
+					},
+				},
+				"matching_target_3": {
+					FilterAttributes: map[string]string{
+						"type": eventType,
+					},
+				},
+			},
+			hasTrigger: true,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			ctx := logtest.TestContextWithLogger(t)
+			psSrv := pstest.NewServer()
+			defer psSrv.Close()
+			psClient := createPubsubClient(ctx, t, psSrv)
+
+			testBrokerConfig := &config.TargetsConfig{
+				Brokers: map[string]*config.Broker{
+					"test_ns_1/test_broker_1": {
+						DecoupleQueue: DecoupleQueue,
+						Targets:       test.brokerTargets,
+					},
+				},
+			}
+
+			brokerConfig := memory.NewTargets(testBrokerConfig)
+			sink := NewMultiTopicDecoupleSink(ctx, brokerConfig, psClient, pubsub.DefaultPublishSettings)
+
+			event := createTestEvent(uuid.New().String())
+
+			hasTrigger := sink.hasTrigger(ctx, event)
+			if hasTrigger != test.hasTrigger {
+				t.Errorf("Sink says event has trigger %t which should be %t", hasTrigger, test.hasTrigger)
+			}
+		})
+	}
+}
+
+func TestMultiTopicDecoupleSinkSendChecksFilter(t *testing.T) {
+	origEnableEventFilterFunc := EnableEventFilterFunc
+	defer func() { EnableEventFilterFunc = origEnableEventFilterFunc }()
+	EnableEventFilterFunc = func() bool {
+		return true
+	}
+
+	filterCalled := false
+	origEventFilterFunc := EventFilterFunc
+	defer func() { EventFilterFunc = origEventFilterFunc }()
+	EventFilterFunc = func(ctx context.Context, attrs map[string]string, event *event.Event) bool {
```

Comment:
Can we remove the mock and check whether the message made it to the topic or not? Then we test that `EventFilterFunc` is set correctly in the tested code. (Arguably E2E tests would also cover this, but having it in a unit test makes identifying a bug much quicker)

Reply:
This change would potentially make e2e test harder for filtering in fanout/retry. Today, we just need to send an event that doesn't match any trigger and expect not receiving it. We know for sure it's the filtering in fanout/retry makes it (not) happen. But with this change, the event would be filtered on ingress.

Reply:
I decided to keep the mock. If we remove it, testing the case where an event shouldn't make it down stream is difficult with current pub/sub. It would make the tests overly complicated imo. so i didn't spend any more time on it. Hope that's okay